[Glimmer2] inline if

GavinJoyce opened the pull request:
part of the [Glimmer Big-Picture Integration Checklist](https://github.com/emberjs/ember.js/issues/12907)

TODO:
- [x] additional tests
  - [x] assertions
  - [x] https://github.com/emberjs/ember.js/pull/12914#issuecomment-180194303
    - [x] fixme: https://github.com/emberjs/ember.js/issues/12925 **note: we're not going to fix this, the tests are running and passing for `@glimmer` only**
  - [x] `BASIC_TRUTHY_TESTS` & `BASIC_FALSY_TESTS`
- [x] refactor `SharedConditionalsTest` so that it can be used by both block and inline helpers
  - [x] extract specialized classes
- [x] rebase after https://github.com/emberjs/ember.js/pull/12919 has landed
- [x] rebase after master is :green_apple: https://github.com/emberjs/ember.js/pull/12919#issuecomment-181028742
- [x] rebase after https://github.com/emberjs/ember.js/pull/12924
- [x] docs
- [x] add support for module `@glimmer` or `@htmlbars` prefixes
- [x] figure out why these are failing: https://github.com/emberjs/ember.js/pull/12920#issuecomment-182658028 (**spoiler:** [jQuery caches](https://github.com/emberjs/ember.js/pull/12920#issuecomment-182908742 data attributes))
- [x] remove htmlbars tests which are now covered in the new shared tests
- [x] move remaining htmlbars inline if tests over to shared tests

GavinJoyce committed:
Glimmer2 inline if

mixonic commented on packages/ember-glimmer/lib/helpers/inline-if.js:
```diff
@@ -0,0 +1,11 @@
+import { toBool as emberToBool } from './if-unless';
+
+//TODO: GJ: docs
```
@rwjblue says the [docs for if](https://github.com/emberjs/ember.js/blob/master/packages/ember-htmlbars/lib/helpers/if_unless.js#L9) should be moved to this file.

mixonic replied:
And thanks for noting this. In 1.13 we lost a ton of docs b/c PRs were merged without this stuff. Definitely worth doing up front.

GavinJoyce replied:
thanks

GavinJoyce commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,467 @@
+import ArrayProxy from 'ember-runtime/system/array_proxy';
+
+//TEMP: GJ: extract and reuse from elsewhere (these are moving in https://github.com/emberjs/ember.js/pull/12919)
+class AbstractConditionalsTest extends RenderingTest {
```
Temp: I'll reuse `AbstractConditionalsTest` once https://github.com/emberjs/ember.js/pull/12919 has merged

rwjblue replied:
#12919 is landed

GavinJoyce commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,467 @@
+
+    //TODO: GJ: update common `AbstractConditionalsTest` to use this format
+    for (let i = 1; i <= values.length; i++) {
+      templates.push(this.templateFor({ cond: `cond${i}`, truthy: `t${i}`, falsy: `f${i}` }));
```
with a slight change, we could reuse this logic for both syntax and inline conditional tests. 

Syntax:
`templates.push(this.templateFor({ cond: `cond${i}`, truthy: `{{t${i}}}`, falsy: `{{f${i}}}` }));`

Inline:
`templates.push(this.templateFor({ cond: `cond${i}`, truthy: `t${i}`, falsy: `f${i}` }));`

Perhaps we should define `AbstractSyntaxConditionalsTest` and `AbstractInlineConditionalsTest`

homu commented:
:umbrella: The latest upstream changes (presumably #12924) made this pull request unmergeable. Please resolve the merge conflicts.

GavinJoyce commented:
I'm going to work on this later tonight

GavinJoyce commented:
There are a number of shared tests which fail in `htmlbars`. For example, 6 of these fail ([related issue](https://github.com/emberjs/ember.js/issues/12925)):

```javascript
moduleFor('Helpers test: {{if}} used with another helper', class extends SharedHelperConditionalsTest {
  templateFor({ cond, truthy, falsy }) {
    return `{{concat (if ${cond} ${truthy} ${falsy})}}`;
  }
});
```

I want to avoid marking these shared tests with a `@glimmer` prefix as that would result in them not being run for `htmlbars` for all other modules. Instead, I'm planning to add support for `@glimmer` and `@htmlbars` module prefixes. This would allow us skip these tests in `htmlbars`:

```javascript
moduleFor('@glimmer Helpers test: {{if}} used with another helper', class extends SharedHelperConditionalsTest {
  templateFor({ cond, truthy, falsy }) {
    return `{{concat (if ${cond} ${truthy} ${falsy})}}`;
  }
});
```

GavinJoyce commented:
I'm seeing 14 test failures for both `@htmlbars` and `@glimmer` for the following test:

```
moduleFor('@glimmer Helpers test: {{if}} used in attribute position', class extends SharedHelperConditionalsTest {

  templateFor({ cond, truthy, falsy }) {
    return `<div data-foo="{{if ${cond} ${truthy} ${falsy}}}" />`;
  }

  textValue() {
    return jQuery('div', this.element).toArray().map(
      function(el) { return jQuery(el).data('foo'); }
    ).join('');
  }

});
```

an example failing test is `[glimmer] Helpers test: {{if}} used in attribute position: it considers empty arrays falsy`. Do we expect these to pass for attributes? I'm going to take try create a failing test in glimmer2 now

chancancode commented:
@GavinJoyce I think the problem is you didn't quote the "truthy" and "falsy" properly: it currently something like `<div data-foo="{{if cond1 T1 F1}}" />` (which looks up `T1` and `F1` as properties on `self`) instead of `<div data-foo="{{if cond1 'T1' 'F1'}}" />`.

By the way, I *think* it will be better to add `wrapperFor` so that we can generate `<div data-foo="{{if cond1 'T1' 'F1'}}{{if cond2 'T2' 'F2'}}{{if cond3 'T3' 'F3'}}" />` instead of `<div data-foo="{{if cond1 'T1' 'F1'}}" /><div data-foo="{{if cond2 'T2' 'F2'}}" /><div data-foo="{{if cond3 'T3' 'F3'}}" />`.

GavinJoyce commented:
@chancancode thanks, I'll implement the `wrapperFor` as you suggest.

I don't think it is a quoting problem, the `renderValues` function that I defined in `SharedHelperConditionalsTest` generates a template and context as follows:

`<div data-foo="{{if cond1 t1 f1}}" /><div data-foo="{{if cond2 t2 f2}}" />`
`{"t1":"T1","f1":"F1","cond1":["hello"],"t2":"T2","f2":"F2","cond2":[]}`

I can't immediately see why this isn't working, but I'm pretty sure I'm doing something wrong. I'll change it to use `wrapperFor`, the problem will likely go away.

chancancode commented:
Ah, you are right! I forgot that we changed it to be dynamic

GavinJoyce commented:
TODO: figure out why this passes:

```javascript
moduleFor('Helpers test: {{if}} used in attribute position', class extends SharedHelperConditionalsTest {
  wrapperFor(templates) {
    return `<div>${templates.join('')}</div>`;
  }

  templateFor({ cond, truthy, falsy }) {
    return `{{if ${cond} ${truthy} ${falsy}}}`;
  }

  textValue() {
    return jQuery('div', this.element).text();
  }
});
```

but this doesn't:

```javascript
moduleFor('Helpers test: {{if}} used in attribute position', class extends SharedHelperConditionalsTest {
  wrapperFor(templates) {
    return `<div data-foo="${templates.join('')}" />`;
  }

  templateFor({ cond, truthy, falsy }) {
    return `{{if ${cond} ${truthy} ${falsy}}}`;
  }

  textValue() {
    return jQuery('div', this.element).data('foo');
  }
});
```

GavinJoyce commented:
@paddyobrien solved this for me. It turns out that jQuery caches the data attribute values when using:

`jQuery('div', this.element).data('foo');` 

This solves it:

`jQuery('div', this.element).attr('data-foo');`

chancancode commented:
Ahh I see. It's probably good to use attr anyway so jQuery doesn't try to parse the text

GavinJoyce commented:
This is ready for review

rwjblue commented:
@chancancode - Would you mind reviewing?

chancancode commented:
@GavinJoyce can you do me a favor and annotate the deleted test case with comments like these? https://github.com/emberjs/ember.js/pull/12914#discussion_r52056719 We are trying to be careful to not lose coverage as we migrate these, and having those annotations would make it a lot easier to review! :heart: 

The tests doesn't always have to be a direct port. We implicitly test a lot more things with the new harness/pattern, so it is quite possible we absorbed some existing test cases into other test cases implicitly (e.g. we used to have separate test cases for re-render, but we now test re-render all the time, so it makes sense to just delete it).

GavinJoyce commented:
@chancancode sure, I'll do that now

GavinJoyce commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,63 @@
+
+}, BASIC_TRUTHY_TESTS, BASIC_FALSY_TESTS);
+
+moduleFor('@glimmer Helpers test: {{if}} used with another helper', class extends SharedHelperConditionalsTest {
```
`@glimmer` only as several of these tests fail in `@htmlbars` (see https://github.com/emberjs/ember.js/issues/12925)

chancancode replied:
:cry:

GavinJoyce commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,63 @@
+
+});
+
+moduleFor('@glimmer Helpers test: {{if}} used in attribute position', class extends SharedHelperConditionalsTest {
```
@glimmer only as several of these tests fail in @htmlbars

chancancode replied:
:cry:

chancancode replied:
:cry:

GavinJoyce commented on packages/ember-glimmer/tests/utils/shared-conditional-tests.js:
```diff
@@ -374,6 +384,53 @@ export class SharedConditionalsTest extends AbstractConditionalsTest {
+}
+
+export class SharedSyntaxConditionalsTest extends SharedConditionalsTest {
   ['@htmlbars it does not update when the unbound helper is used']() {
```
Moved to `SharedSyntaxConditionalsTest` as this template isn't compatible with `SharedHelperConditionalsTest`. (Having said that, I'm going to reimplement a `SharedHelperConditionalsTest` version of it now)

GavinJoyce commented:
@chancancode I've added a `SharedHelperConditionalsTest` version of the `@htmlbars it does not update when the unbound helper is used` test, as it needs a different type of template.

I've also annotations for the two modules which are `@glimmer` only. I haven't removed any tests other tests, the diff only looks like it due to a few chunks of content moving around.

GavinJoyce commented on packages/ember-glimmer/tests/utils/abstract-test-case.js:
```diff
@@ -16,7 +16,11 @@ const packageTag = `@${packageName} `;
-  QUnit.module(`[${packageName}] ${description}`, {
+  let modulePackagePrefixMatch = description.match(/^@(\w*)/); //eg '@glimmer' or '@htmlbars'
+  let modulePackagePrefix = modulePackagePrefixMatch ? modulePackagePrefixMatch[1] : '';
+  let descriptionWithoutPackagePrefix = description.replace(/^@\w* /, '');
```
this adds support for module prefixes such as `@glimmer` and `@htmlbars` and is needed as the comprehensive shared tests have highlighted some issues with htmlbars

GavinJoyce commented on packages/ember-glimmer/tests/utils/shared-conditional-tests.js:
```diff
@@ -374,11 +384,56 @@ export class SharedConditionalsTest extends AbstractConditionalsTest {
+    let context = {};
+
+    for (let i = 1; i <= values.length; i++) {
+      templates.push(this.templateFor({ cond: `cond${i}`, truthy: `t${i}`, falsy: `f${i}` }));
```
we can't use `renderValues` from `SharedConditionalsTest` as it uses 

```
{ cond: `cond${i}`, truthy: `{{t}}${i}`, falsy: `{{f}}${i}` }
```

and that would produce a template such as:

```
<div data-foo="{{if cond1 {{t}}1 {{f}}1}}" />
```

chancancode replied:
I see. So the problem is that we got a little too greedy in the original test case. It was originally written as `{{#if cond1}}T1{{else}}F1{{/if}}`, but then we realized that it's generally important to smoke test that both templates ("default" and "inverse") have access to the ambient scope, so we changed it to be `{{#if cond1}}{{t}}1{{else}}{{f}}1{{/if}}` to test that it has access to `t` and `f` on the outer scope. (We also somehow decided that it is important to test that `{{t}}1` works, as opposed to just `{{t1}}`.)

So... we can do one of two things here:

1. We decide that "obviously" that the condition position and the value positions share the same sccope, and there are no risk of scope switching here, and since the condition is already dynamic (tests accessing the scope), there is no need of testing that in the value positions as well. In which case, we can generate something like `{{if cond1 "T1" "F1"}}`.

2. We maintain the spirit of the original test, and generates `{{if cond1 (concat t "1") (concat f "1")}}`.

3. We go above and beyond and make all three parts dynamic, like you did here.

I think all of these positions are actually quite reasonable, and perhaps it even makes some sense to want to test *all* of them. The problem I see now is that we decided to go with (3) here, but in the actually `inline-if-test.js` we *also* felt the need to have *some* tests for (1) (that's what "static argument", "it can use falsy param as binding", "it respects isTruthy when object changes" etc are about, if I understand correctly).

If we decided to care about multiple of those  scenarios, I think it makes more sense to use the same set of test cases and subclass multiple times with different `renderValues` implementations instead of doing one set of them here and the rest in an ad-hoc fashion in the concrete subclass.

For example, we have a pretty exhaustive sets of `isTruthy` tests in the shared test cases (bare object with `isTruthy`, `Ember.Object` various proxies, testing both true and false as the starting value, etc), but we only have a single row of that matrix in the concrete subclass for (1) at the moment:

```javascript
  ['@test it respects isTruthy when object changes']() {
     this.render(`{{if condition 'truthy' 'falsy'}}`, { condition: EmberObject.create({ isTruthy: false }) });
 
     this.assertText('falsy');
 
     this.runTask(() => this.rerender());
 
     this.assertText('falsy');
 
     this.runTask(() => set(this.context, 'condition', EmberObject.create({ isTruthy: true })));
 
     this.assertText('truthy');
 
     this.runTask(() => set(this.context, 'condition', EmberObject.create({ isTruthy: false })));
 
     this.assertText('falsy');
   }
```

The TL;DR is that I think we should either decide that one of these variations is enough, and remove most of the ad-hoc tests in the subclass, or we subclass this multiple times.

Does that make sense?

GavinJoyce replied:
Thanks, that makes sense.

The `"it can use falsy param as binding"`, `"it respects isTruthy when object changes"``... tests were direct ports from htmlbars, happened after I had added support for (3) and I didn't really give them much thought.

It's late here, but I'll take a higher level view of all these tests tomorrow, experiment a little and report back

GavinJoyce replied:
I've gone through the ad-hoc tests, implemented your comments and removed tests which where covered in the shared cases. There is only one ad-hoc one remaining, `it can omit the falsy argument`. My feeling is that it's fine for us to just test a single variation in this case, the shared test cases are fully exercising the truthy and falsy scenarios and the `it can omit the falsy argument` test verifies the behaviour of when the last parameter is omitted.

chancancode replied:
:+1: :+1: :+1: :+1:

GavinJoyce commented:
Let me know if you'd like me to annotate anything else

chancancode commented:
@GavinJoyce sorry! I should have looked more carefully – I copied the comment from [here](https://github.com/emberjs/ember.js/pull/12927#issuecomment-183062997) assuming it would make sense.

Can you review the existing tests and either delete (with the annotation) them or migrate them into the new test files? If some of them are too hard to port we can leave them in the old file and come back to them later; but I suspect your work already covered the majority of them, so seems good to have you review and delete them while your memory is still fresh.

Great work so far by the way, thank you so much for picking this up!!

GavinJoyce commented:
Ah, thanks for the clarification. Sure, I'll do that on Saturday which will be the next time I'll get to work on this.

I'm happy to continue working on glimmer2 integration once this has merged too, we're very excited to try  it our Intercom app

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -205,67 +205,6 @@ QUnit.test('edge case: rerender appearance of inner virtual view', function() {
   equal(jQuery('#qunit-fixture').text(), 'test');
 });
 
-QUnit.test('`if` helper with inline form: renders the second argument when conditional is truthy', function() {
```
replaced with `it renders the corresponding block based on the conditional`

chancancode replied:
:metal: (this should be covered by the generic tests once we have resolved https://github.com/emberjs/ember.js/pull/12920/files#r53242487)

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -205,67 +205,6 @@ QUnit.test('edge case: rerender appearance of inner virtual view', function() {
-  equal(view.$().text(), 'truthy');
-});
-
-QUnit.test('`if` helper with inline form: renders the third argument when conditional is falsy', function() {
```
//GJ: replaced with `it renders the corresponding block based on the conditional`

chancancode replied:
:metal: (this should be covered by the generic tests once we have resolved https://github.com/emberjs/ember.js/pull/12920/files#r53242487)

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -205,67 +205,6 @@ QUnit.test('edge case: rerender appearance of inner virtual view', function() {
-  equal(view.$().text(), 'falsy');
-});
-
-QUnit.test('`if` helper with inline form: can omit the falsy argument', function() {
```
replaced with `it can omit the falsy argument`

chancancode replied:
:metal:

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -205,67 +205,6 @@ QUnit.test('edge case: rerender appearance of inner virtual view', function() {
-  equal(view.$().text(), '');
-});
-
-QUnit.test('`if` helper with inline form: truthy and falsy arguments are changed if conditional changes', function() {
```
replaced with `it renders the corresponding block based on the conditional`

chancancode replied:
:metal:

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -303,77 +242,6 @@ QUnit.test('`if` helper with inline form: can use falsy param as binding', funct
   equal(view.$().text(), 'no');
 });
 
-QUnit.test('`if` helper with inline form: raises when using more than three arguments', function() {
```
replaced with `it raises when there are more than three arguments`

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -303,77 +242,6 @@ QUnit.test('`if` helper with inline form: can use falsy param as binding', funct
-  }, /The inline form of the `if` and `unless` helpers expect two or three arguments/);
-});
-
-QUnit.test('`if` helper with inline form: raises when using less than two arguments', function() {
```
replaced with `it raises when there are less than two arguments`

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -303,77 +242,6 @@ QUnit.test('`if` helper with inline form: can use falsy param as binding', funct
-  }, /The inline form of the `if` and `unless` helpers expect two or three arguments/);
-});
-
-QUnit.test('`if` helper with inline form: works when used in a sub expression', function() {
```
replaced with module: `Helpers test: {{if}} used with another helper` and test: `it renders the corresponding block based on the conditional`

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -303,77 +242,6 @@ QUnit.test('`if` helper with inline form: can use falsy param as binding', funct
-  equal(view.$().text(), 'truthy');
-});
-
-QUnit.test('`if` helper with inline form: updates if condition changes in a sub expression', function() {
```
replaced with module: `Helpers test: {{if}} used with another helper` and test: `it renders the corresponding block based on the conditional`

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -303,77 +242,6 @@ QUnit.test('`if` helper with inline form: can use falsy param as binding', funct
-  equal(view.$().text(), 'innerFalsy');
-});
-
-QUnit.test('`if` helper with inline form: can use truthy param as binding in a sub expression', function() {
```
replaced with module: `Helpers test: {{if}} used with another helper` and test: `it renders the corresponding block based on the conditional`

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -397,56 +265,6 @@ QUnit.test('`if` helper with inline form: respects isTruthy when object changes'
   equal(view.$().text(), 'falsy');
 });
 
-QUnit.test('`if` helper with inline form: respects isTruthy when property changes', function() {
```
replaced by `it renders the corresponding block based on the conditional`

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -397,56 +265,6 @@ QUnit.test('`if` helper with inline form: respects isTruthy when object changes'
-  equal(view.$().text(), 'falsy');
-});
-
-QUnit.test('`if` helper with inline form: respects length test when list content changes', function() {
```
replaced with `it considers empty arrays falsy`

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -205,67 +205,6 @@ QUnit.test('edge case: rerender appearance of inner virtual view', function() {
-  equal(view.$().text(), 'truthy');
-});
-
-QUnit.test('`if` helper with inline form: can omit the falsy argument and renders nothing when conditional is falsy', function() {
```
replaced with `it can omit the falsy argument`

chancancode replied:
:metal: (see comment in that new test about testing both starting conditions)

GavinJoyce commented:
@chancancode I've removed 12 htmlbars tests and added annotations to this PR. There are 5 inline if htmlbars tests remaining which I'm planning to move over in a subsequent PR

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -284,215 +222,6 @@ QUnit.test('`if` helper with inline form: can use truthy param as binding', func
   equal(view.$().text(), 'yes');
 });
 
-QUnit.test('`if` helper with inline form: can use falsy param as binding', function() {
```
moved to `it can use falsy param as binding`

chancancode replied:
:metal: see https://github.com/emberjs/ember.js/pull/12920/files#r53243030

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -284,215 +222,6 @@ QUnit.test('`if` helper with inline form: can use truthy param as binding', func
-  equal(view.$().text(), 'innerOk');
-});
-
-QUnit.test('`if` helper with inline form: respects isTruthy when object changes', function() {
```
moved to `it respects isTruthy when object changes`

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -284,215 +222,6 @@ QUnit.test('`if` helper with inline form: can use truthy param as binding', func
-  equal(view.$().text(), 'falsy');
-});
-
-QUnit.test('`if` helper with inline form: updates when given a falsey second argument', function() {
```
moved to `it updates when given a falsey second argument`

chancancode replied:
:metal:

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -284,215 +222,6 @@ QUnit.test('`if` helper with inline form: can use truthy param as binding', func
-  equal(view.$().text(), 'falsy');
-});
-
-QUnit.test('`if` helper with inline form: respects length test when list itself', function() {
```
replaced with `it respects length test when list itself`

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -284,215 +222,6 @@ QUnit.test('`if` helper with inline form: can use truthy param as binding', func
-  equal(view.$().text(), 'falsy');
-});
-
-QUnit.test('`if` helper with inline form: respects isTruthy when property changes', function() {
```
replaced with `it respects isTruthy when object changes`

GavinJoyce commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -284,215 +222,6 @@ QUnit.test('`if` helper with inline form: can use truthy param as binding', func
-  equal(view.$().text(), 'no');
-});
-
-QUnit.test('`if` helper with inline form: raises when using more than three arguments', function() {
```
replaced with `it raises when there are more than three arguments`

GavinJoyce replied:
I've just realised that hasn't been moved over fully as the shared test uses the `@glimmer` prefix due to the assertion message being different. I'll update the shared assertion messages now so that we don't lose coverage

GavinJoyce replied:
done

chancancode replied:
Great catch. It seems a bit annoying that we have to disable the entire module just because a few tests fail, but I don't know what's the best way to fix it.

GavinJoyce commented:
I've moved all remaining htmlbars inline if tests across to the shared tests and have annotated each with comments on the PR.

rwjblue commented:
Awesome, thank you @GavinJoyce!

@chancancode - r?

GavinJoyce commented:
I just made a small update as I noticed that the new assertion tests we're prefixed with `@glimmer`, they are now prefixed with `@test`

chancancode commented:
Thanks for your work on this! We finally finished reviewing (and merged) #12927 and will do this next! (Is there anything we need to update to account for the changes in #12927?)

GavinJoyce commented:
@chancancode I'll take a look at the comments on https://github.com/emberjs/ember.js/pull/12927 now and I'll report back soon

GavinJoyce commented:
@chancancode I reviewed the comments and made on small set of changes to the ported tests so that they follow the `render -> rerender -> mutate -> render pattern`. This is ready for review

GavinJoyce commented:
The build is failing after rebasing, I'm looking into it now

GavinJoyce commented on packages/ember-glimmer/tests/utils/shared-conditional-tests.js:
```diff
@@ -374,11 +384,56 @@ export class SharedConditionalsTest extends AbstractConditionalsTest {
   }
 
+  ['@test it maintains DOM stability when condition changes from a truthy to a different truthy value']() {
+    this.renderValues(true);
```
this was moved without change

chancancode replied:
:+1: :+1: :+1: :+1: :+1:

GavinJoyce commented on packages/ember-glimmer/tests/utils/shared-conditional-tests.js:
```diff
@@ -374,11 +384,56 @@ export class SharedConditionalsTest extends AbstractConditionalsTest {
+    this.assertInvariants();
+  }
+
+  ['@test it maintains DOM stability when condition changes from a falsy to a different falsy value']() {
```
this was moved without change

chancancode replied:
:+1: :+1: :+1: :+1: :+1:

GavinJoyce commented:
The build is mostly :green_apple: now, the first job failed due to `ERROR 429: TOO MANY REQUESTS`.

This is ready for review (really this time!)

chancancode commented on packages/ember-glimmer/lib/environment.js:
```diff
@@ -51,9 +51,11 @@ class PropertyReference {
 const helpers = {
-  concat
+  concat,
+  if: inlineIf
```
We have an `inlineUnless` helper in Ember too :smile:

GavinJoyce replied:
I can quickly follow this PR up with one adding support for inline unless
if that's OK?
On 17 Feb 2016 6:47 p.m., "Godfrey Chan" <notifications@github.com> wrote:

> In packages/ember-glimmer/lib/environment.js
> <https://github.com/emberjs/ember.js/pull/12920#discussion_r53210816>:
>
> >
> >  const helpers = {
> > -  concat
> > +  concat,
> > +  if: inlineIf
>
> We have an inlineUnless helper in Ember too [image: :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/emberjs/ember.js/pull/12920/files#r53210816>.
>

chancancode replied:
:+1:

chancancode commented on packages/ember-glimmer/lib/helpers/inline-if.js:
```diff
@@ -0,0 +1,38 @@
+    return args[1];
+  } else {
+    let falsyArgument = args[2];
+    return falsyArgument === undefined ? '' : falsyArgument;
```
I suppose this is needed to pass the tests at the moment? Can you add a TODO here? I believe we should unconditional return `args[2]` here (even if it's undefined) and let a different part of the system handle converting `undefined`, `null`, etc to `""` (which we haven't implemented yet).

It's probably good to add a test for this case if we don't already have one (I think it would have to be `@htmlbars` for now), but perhaps that should be part of the general [checklist item](https://github.com/emberjs/ember.js/issues/12907) of "Add tests for rending false, undefined, null etc in content-test.js, similar to..."

```javascript
this.render(`{{if condition truthy}}`, { condition: true, truthy: null });
this.render(`{{if condition truthy}}`, { condition: true, truthy: undefined });
this.render(`{{if condition truthy}}`, { condition: true, truthy: false });
...etc
```

GavinJoyce replied:
yes, this was added to allow the moved `it can omit the falsy argument` test to pass. I'll modify this test to be `@htmlbars` and add a TODO as you suggest.

chancancode replied:
Seems like this got lost in the merge too, we retained the TODO comment but lost the @htmlbars marker

chancancode commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,150 @@
+    this.runTask(() => set(this.context, 'condition', false));
+
+    this.assertText('no');
+  }
```
Isn't this the same as the `BASIC_TRUTHY_TESTS`? (The condition isn't really static)

chancancode replied:
A more general point (even though I don't think we need this test): the general pattern we have been following is:

1. Initial render
2. No-op re-render
3. Change (via mutation if applicable)
4. Repeat (3) as needed (sometimes with replacement, or other values, sometimes grouped, ...)
5. Reset (via replacement if applicable)

In this case, you would change render it with `true` initially, then `rerender` with no change, then change it to `false`, and finally change it back to `true` (replacement vs mutation doesn't make a difference here).

See https://github.com/emberjs/ember.js/pull/12927#discussion_r52392940 for more details

GavinJoyce replied:
> Isn't this the same as the BASIC_TRUTHY_TESTS? (The condition isn't really static)

I'm not sure that the `BASIC_TRUTHY_TESTS` cover this, they currently drive templates such as `{{if ${cond} ${truthy} ${falsy}}}` but not `{{if ${cond} "T" "F"}}`. This behaviour is covered by tests such as `it respects isTruthy when object changes` though, so I'll drop this test as you suggest.

> the general pattern we have been following is: .... 5. Reset (via replacement if applicable)`

I'll update the tests to include step 5 which I missed when reviewing the `with` PR

chancancode replied:
Hmm, good point!

GavinJoyce commented:
@chancancode thanks for your feedback. I've made the following changes:

 * added comment : https://github.com/emberjs/ember.js/pull/12920/files#diff-8bf5a47bad15668bbc3a067fefd4118dR35
 * made related test `@htmlbars`: https://github.com/emberjs/ember.js/pull/12920/files#diff-034be756e33abf234e1e3ff8f648e810R18
 * removed the `it allows static arguments` test which was covered elsewhere
 * ensured that all remaining tests have the `Reset (via replacement if applicable)` step

chancancode commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,144 @@
+    this.runTask(() => set(this.context, 'condition', true));
+
+    this.assertText('yes');
+  }
```
:+1: 

We should test both starting conditions (starting with `true` and starting with `false`), like this: https://github.com/emberjs/ember.js/blob/master/packages/ember-glimmer/tests/integration/syntax/if-unless-test.js#L52-L79

GavinJoyce replied:
Thanks, done

chancancode commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  ['@test it can use falsy param as binding']() {
+    this.render(`{{if condition truthy falsy}}`, { condition: false, truthy: 'yes', falsy: 'no' });
```
I think this is the "start with `false`" version of the "static argument" test? In general, we try to group them in a single test (see the comment above) so that we don't have an explosion of test cases. But this is also related to the broader point of https://github.com/emberjs/ember.js/pull/12920/files#r53242487

chancancode commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  ['@test it respects isTruthy when object changes']() {
+    this.render(`{{if condition 'truthy' 'falsy'}}`, { condition: EmberObject.create({ isTruthy: false }) });
```
See https://github.com/emberjs/ember.js/pull/12920/files#r53242487

GavinJoyce replied:
This is covered by `it tests for `isTruthy` on Ember objects if available` so I'll remove it

chancancode commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  ['@test it respects length test when list itself']() {
+    this.render(`{{if condition 'truthy' 'falsy'}}`, { condition: [] });
```
See https://github.com/emberjs/ember.js/pull/12920/files#r53242487

chancancode commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,144 @@
+
+    this.assertText('falsy');
+
+    this.runTask(() => set(this.context, 'condition', [1]));
```
I don't think we will end up keeping this (see above), but FYI this is a case where we would generally test `mutation` vs `replacement` (although we will have to use Ember.A and `pushObject`): https://github.com/emberjs/ember.js/blob/master/packages/ember-glimmer/tests/utils/shared-conditional-tests.js#L244-L273

GavinJoyce replied:
:+1:

chancancode commented on packages/ember-glimmer/tests/utils/shared-conditional-tests.js:
```diff
@@ -154,6 +151,19 @@ export const BASIC_FALSY_TESTS = {
 
 export class SharedConditionalsTest extends AbstractConditionalsTest {
```
What is the difference between `SharedConditionalsTest` and `AbstractConditionalsTest`?

chancancode replied:
I see. Turns out I wrote that code and I didn't even remember/understand it :P `AbstractConditionalsTest` is like an "interface" for the most part and `SharedConditionalsTest` has the test case implementations. Presumably, I did that for clearity but I'm not sure if it end up helping :P

But in any case, I think this supports my comment below that the `renderValues` implementation should be pushed one level down in the class hierarchy.

GavinJoyce replied:
:+1:

chancancode commented on packages/ember-glimmer/tests/utils/shared-conditional-tests.js:
```diff
@@ -154,6 +151,19 @@ export const BASIC_FALSY_TESTS = {
+
+    let wrappedTemplate = this.wrapperFor(templates);
+    this.render(wrappedTemplate, assign({ t: 'T', f: 'F' }, context));
+  }
```
Isn't this implementation specific to the syntax version? (because it uses `{{t}}...`) Shouldn't this be in `SharedSyntaxConditionalsTest`? (it might be related to me not understanding the difference between the class hierarchy)

GavinJoyce replied:
:+1:

chancancode commented on packages/ember-glimmer/tests/utils/shared-conditional-tests.js:
```diff
@@ -407,33 +462,41 @@ export class SharedConditionalsTest extends AbstractConditionalsTest {
-
-    this.assertText('T1');
+export class SharedSyntaxConditionalsTest extends SharedConditionalsTest {
+  ['@htmlbars it does not update when the unbound helper is used']() {
```
Not to self: this is `@htmlbars` for now because we haven't implemented `(unbound` for glimmer. I wish we had real method decorators so we could write something like `@skip("glimmer", "unbound is not yet implemented")`. Maybe we should make it a rule for now that we should add a comment above each `@htmlbars` and `@glimmer`?

chancancode replied:
@rwjblue just curious, how difficult would it be to write a babel plugin that lets us write something like this?

```javascript
  // @test
  ["test name"] {
    ...
  }

  // @test @skip("glimmer", "unbound is not yet implemented")
  ["another test"] {
    ...
  }
```

(But maybe at that point we are really just implementing method decorators...)

chancancode commented on packages/ember-glimmer/tests/utils/shared-conditional-tests.js:
```diff
@@ -407,33 +462,41 @@ export class SharedConditionalsTest extends AbstractConditionalsTest {
+    this.assertText('T1F2');
   }
 
   ['@test it tests for `isTruthy` on the context if available']() {
```
We should probably have a version for this for the helper version too. Can you add that?

While you are at that, I noticed this test doesn't test both starting conditions. Can you try to copy the format of the test immediately above us to test both starting conditions in parallel?

GavinJoyce replied:
I'm not quite sure how I can test both starting conditions without creating two distinct tests as the condition is `this` and there is only one of those.

chancancode commented on packages/ember-htmlbars/tests/helpers/if_unless_test.js:
```diff
@@ -284,215 +222,6 @@ QUnit.test('`if` helper with inline form: can use truthy param as binding', func
   equal(view.$().text(), 'yes');
 });
```
I *think* the spirit of this test is just to test `{{if dynamicCond dynamicTruthy}}`, which should be more or less covered by your new setup. The differece between this and the new `@htmlbars it can omit the falsy argument` is that the second argument is dynamic, which falls into the bucket of https://github.com/emberjs/ember.js/pull/12920/files#r53242487. If we decided that it doesn't matter, we can just delete this. If we think it matters, we should test that in both configurations (and then delete this).

GavinJoyce replied:
I'm happy to drop this. Let me know if you object, otherwise I'll delete it before merging

TODO: 

 * [ ] Either delete this or rewrite it in shared conditionals

chancancode replied:
Oops, I didn't notice this

GavinJoyce replied:
I'll follow up on this with a PR this weekend

chancancode commented:
I ran out of time for this review session, I'll return to it another time! But I have left some "big picture" comments for you to chew on at the moment!

GavinJoyce commented:
Thanks for your feedback @chancancode. Like you, I'm keen to nail these patterns early so that the we can quickly and comprehensively integrate the rest of glimmer2 into ember

chancancode commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,144 @@
+    this.assertText('falsy');
+  }
+
+  ['@test it updates when given a falsey second argument']() {
```
:metal: (I guess this must be a regression test?)

chancancode commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,144 @@
+
+  templateFor({ cond, truthy, falsy }) {
+    return `{{concat (if ${cond} ${truthy} ${falsy})}}`;
+  }
```
This should probably be

```javascript
  templateFor({ cond, truthy, falsy }) {
    return `(if ${cond} ${truthy} ${falsy})`;
  }

  wrapperFor(templates) {
    return `{{concat ${templates.join(' ')}}}`;
  }
```

GavinJoyce replied:
thanks, done

chancancode commented on packages/ember-glimmer/tests/integration/helpers/inline-if-test.js:
```diff
@@ -0,0 +1,144 @@
+
+  textValue() {
+    return jQuery('div', this.element).attr('data-foo');
+  }
```
We can make a `this.$('div')` :smile:

GavinJoyce replied:
thanks, done

GavinJoyce replied:
FYI, I added `this.$ = jQuery;` to `RenderingTest` as `this.$` wasn't already available

GavinJoyce commented on packages/ember-glimmer/tests/utils/shared-conditional-tests.js:
```diff
@@ -55,6 +52,7 @@ export const BASIC_TRUTHY_TESTS = {
     { foo: 'bar' },
     EmberObject.create(),
     EmberObject.create({ foo: 'bar' }),
+    EmberObject.create({ isTruthy: true }),
```
I added this as it seemed like an omission

chancancode replied:
:metal:

GavinJoyce commented on packages/ember-glimmer/tests/utils/shared-conditional-tests.js:
```diff
@@ -99,7 +97,8 @@ export const BASIC_FALSY_TESTS = {
     [],
-    emberA()
+    emberA(),
+    EmberObject.create({ isTruthy: false })
```
I added this as it seemed like an omission

chancancode replied:
:metal:

chancancode commented on packages/ember-glimmer/tests/utils/abstract-test-case.js:
```diff
@@ -78,6 +86,7 @@ export class RenderingTest extends TestCase {
     let owner = this.owner = buildOwner();
     let env = this.env = new Environment({ dom, owner });
     this.renderer = new Renderer(dom, { destinedForDOM: true, env });
+    this.$ = jQuery;
```
ah, this is not what I meant, I mean something like how `this.$()` works in components, but I can fix it

chancancode commented:
This is really awesome! Thanks for taking the time to work on this @GavinJoyce :heart: :green_heart: :yellow_heart: :blue_heart: :purple_heart:
